### PR TITLE
Add Qwant search engine

### PR DESCRIPTION
**Pull request by imagoiq**
Hi,

I've started the work to add Qwant search engine. The action "Block this site" is appearing with more or less good style and at the right place, but I'm a bit stuck, because currently it is unreliable with the standard version of Qwant (https://www.qwant.com). The lite version (lite.qwant.com) works perfectly as it's rather simple, but on the normal version the action only appears when we refresh the page.

Maybe @iorate you can clarify the use of `pagerHandlers` and `controlHandlers`. I probably write something wrong there. I would be glad if anyone else wants to help on this PR.

What doesn't work:
* When we navigate between Web, news, images, videos, the action doesn't appear
* When searching from the homepage, the action doesn't appear
* When clicking on "More results" button, the action doesn't appear

### CODE REVIEW

File: website/docs/advanced-features.md [hunk 1]
```diff
@@ -92,6 +92,9 @@ For now, search scopes other than "Web" (e.g. "Images" and "Videos") are not sup
 
 ![ecosia](/img/advanced-features/ecosia.png)
 
+### Qwant
+
+
```
Comment (imagoiq) [inline website/docs/advanced-features.md:97]: Todo

**Comment by iorate** [comment]
Don't worry about lint errors; you can fix them later by running `yarn fix`.

**Comment by iorate** [comment]
Your PR does not work fully in my environment; controls ("uBlacklist has blocked %d sites") and actions ("Block this site") don't show up.
It is because Qwant is built on React and the runtime of React removes controls and actions on re-rendering DOM. In general, it is very diffucult to manage DOM of sites built on virtual DOM frameworks.
I guess you tried to work around this problem by `onSerpElementDelay`, but it does not work intendedly in my environment. The reason may be that I access Qwant via a proxy in the US now (Qwant is not accessible from Japan, where I live), and so the latency of rendering differs from that in your environment.

**Comment by iorate** [comment]
Ecosia support has the same problem because it uses Nuxt.js.
It works only because Ecosia mercifully overlooks actions inserted into the following position:

https://github.com/iorate/ublacklist/blob/ae6031f3fb35cc54a772371b174f5b168381a408/src/scripts/search-engines/ecosia.ts#L34

The control is inserted directly into the `body`.

**Comment by iorate** [comment]
[GHHbD](https://greasyfork.org/ja/scripts/1682-google-hit-hider-by-domain-search-filter-block-sites) and [HoHSER](https://github.com/pistom/hohser) seems to have the same problem.

**Comment by iorate** [comment]
@imagoiq Maybe you could focus on the lite version for now.

**Comment by imagoiq** [comment]
Thanks, I'm still confusing with `pagerHandlers` and `controlHandlers` usage. I took one hour today to try a bit deeper and found out that the MutationObserver only find some changes on some parents levels, so this works:

```
      {
        scope: 'web',
        target:
          '[class*=SearchLayout-module__content] > [class*=Web-module__container], section > [class*=Stack-module__VerticalStack]',
        url: 'a',
        title: 'a',
        actionTarget: '[class*=WebResult-module__permalink]',
        actionStyle: {
          fontSize: '14px',
          paddingLeft: '5px',
        },
      },
```

…but of course only for the first element of the page. Is there any mechanism to trigger entry check on some parents elements? Code is quite hard to follow, so I wouldn't be surprise if I miss something.

**Comment by iorate** [comment]
It is true that the current code is very complicated and I may need to create a document for contributors. For now, let me explain control handlers and pager handlers here.

## Control Handlers
'Control' means an element showing 'uBlacklist has blocked %d sites Show'.
```TypeScript
export type ControlHandler = {
  scope?: string;
  target: string | ((element: HTMLElement) => boolean);
  position?:
    | 'beforebegin'
    | 'afterbegin'
    | 'beforeend'
    | 'afterend'
    | ((target: HTMLElement) => HTMLElement | null);
  style?: string | CSSAttribute | ((root: HTMLElement) => void);
  buttonStyle?: string | CSSAttribute | ((button: HTMLElement) => void);
};
```
- **scope**: A string representing the scope (e.g. 'web' and 'news') of results that the control handles. DuckDuckGo and Qwant are implemented as single page application (SPA). In other words, multiple scopes may exist together in a single page. We should distinguish results of a scope from those of another.
- **target**: A string or function selecting the element where the control should be inserted from elements found by `MutationObserver`. It is usually designated by a CSS selector.
- **position**: A string representing the insertion position relative to `target`.
- **style**: A CSS style applied to the control.
- **buttonStyle**: A CSS style applied to the buttons ('Show' and 'Hide'). In most cases, `globalStyle` should be used instead.

## Pager Handlers
As you pointed out, `MutaionObserver` does not detect all elements. There are two reasons:
- Obviously it cannot detect elements already inserted when the script is injected. By default, however, `handleSerp` deals with this problem by calling `querySelectorAll` on injection and passing the resuls to control & entry handlers.
- It cannot detect children of elements programatically inserted by JavaScript. For example, if you push the 'Display more results' button, Qwant will add an element with a class starting with `Stack-module__VerticalStack` but its children won't be detected by `MutationObserver`. Pager handlers 'salvage' such children and pass them to control & entry handlers.
```TypeScript
export type PagerHandler = {
  target: string | ((element: HTMLElement) => boolean);
  innerTargets: string | ((target: HTMLElement) => HTMLElement[]);
};
```
- **target**: A string or function selecting the element programatically inserted.
- **innerTargets**: A string of function selecting the children of `target` that should be passed to control & entry handlers.

**Comment by iorate** [comment]
There remains a major problem: actions will be removed during the hydraion process of React. One solution could be to wait for the hydration process to exit (for example, by listening to the `load` event of the script that does hydration).

**Comment by imagoiq** [comment]
Thanks. I manage to get a working version on Firefox. On Brave, it seems that it doesn't work on refresh the page, your idea to wait for the hydration process to exit will probably be needed, but I don't know how to do that. 

Maybe we could release a first version targeted to Firefox users? @sherguin, I noticed that you were interested in the support of Qwant, perhaps you could give a try?

File: src/scripts/content-script.tsx [hunk 2]
```diff
@@ -118,12 +118,14 @@ class ContentScript {
...
-            console.debug(addedNode.cloneNode(true));
+            console.debug(addedNode, addedNode.cloneNode(true));
             // #endif
             if (addedNode === document.head && this.options) {
               this.onSerpHead();
             }
-            this.onSerpElement(addedNode);
+            setTimeout(() => {
+              this.onSerpElement(addedNode);
+            }, 0);
```
Comment (imagoiq) [inline src/scripts/content-script.tsx:128]: This is needed to work on refresh in scope "images" on Firefox

**Comment by iorate** [comment]
> your idea to wait for the hydration process to exit will probably be needed, but I don't know how to do that.

I came up with a compromise. Would you try to set this value to `document_idle`? It will delay script injection.
https://github.com/iorate/ublacklist/blob/54677dbf5ccc51eca116bf9020bc36f62dc2b08b/src/scripts/background/search-engines.ts#L21

**Comment by imagoiq** [comment]
Great this works now flawlessly on Chrome and it fix as well the issue on Firefox with "images" scope!

**Comment by iorate** [comment]
I'm glad that it works.

We still have two things to do.

1. Make controls work. The current control handler never works, because it has no scope (defaulted to `''`) while the entry handlers have scopes.
   - (Solution 1) Remove the scopes of the entry handlers to make one control deal with blocked entries of all scopes.
   - (Solution 2) Add a control handler for each scope. In this case, each control should be appropriately placed for each scope. Let me give you an example, the news search of DuckDuckGo.<br>https://github.com/iorate/ublacklist/blob/c87adf38a678c6988f0ff104bc2af8075415098f/src/scripts/search-engines/duckduckgo.ts#L100-L110 ddg-news-control<br>Now that hydration is avoided using the `document_idle` hack, you can probably place controls anywhere, not only as a child of `body`.

1. The use of `document_idle` should be restricted to the standard version of Qwant. The original value `document_start` enables one of the most important features of this extension: "Prevent blocked sites from appearing *even in a moment*". `document_idle` is an emergency hatch for this very case.
    - I will do this after merging this PR, although of course you can do it.

**Comment by imagoiq** [comment]
Great, control handlers are ✅

I'll let you do the to-do number 2 as you might have a better idea where you want to place the check.

**Comment by iorate** [comment]
Seems good to me. Let me review the details.

> I'll let you do the to-do number 2 

No problem.

**Comment by fabform** [comment]
Handy alternative list
[alternative search engine list](https://fabform.io/a/alternative-search-engines)

File: README.md [hunk 3]
```diff
@@ -29,6 +29,8 @@ This extension is available in the below search engines.
 | DuckDuckGo (iOS) | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: |
 | Ecosia           | :heavy_check_mark: |                    |                    |                    |
 | Ecosia (iOS)     | :heavy_check_mark: |                    |                    |                    |
+| Qwant            | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: |
```
Comment (iorate) [inline README.md:32]: Would you add an annotation about the "Always play videos on Qwant.com" option of the video search?
For example, replace the 3rd `:heavy_check_mark:` with `*1` and add a footnote: `*1 Works only if you turn off "Always play videos on Qwant.com".`.

File: README.md [hunk 4]
```diff
@@ -29,6 +29,8 @@ This extension is available in the below search engines.
...
+| Qwant (iOS)      | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: |
```
Comment (iorate) [inline README.md:33]: The same as above

**Comment by fabform** [comment]
Handy alternative list of [alternative search engines](https://fabform.io/a/alternative-search-engines)

File: src/common/search-engines.ts [hunk 5]
```diff
@@ -219,6 +219,7 @@ export const searchEngineMatches: Readonly<Record<SearchEngineId, string[]>> = {
     '*://start.duckduckgo.com/?*',
   ],
   ecosia: ['https://www.ecosia.org/search?*'],
+  qwant: ['https://www.qwant.com/*', 'https://lite.qwant.com/*'],
```
Comment (iorate) [inline src/common/search-engines.ts:222]: Could you limit the match patterns a bit more?
`['https://www.qwant.com/?*', 'https://lite.qwant.com/?*']` (which paths are `/?*`) seems sufficient to me.

File: src/scripts/background/search-engines.ts [hunk 6]
```diff
@@ -18,7 +18,7 @@ export async function injectContentScript(tabId: number, url: string): Promise<v
   if (!active) {
     await apis.tabs.executeScript(tabId, {
       file: '/scripts/content-script.js',
-      runAt: 'document_start',
+      runAt: 'document_idle',
```
Comment (iorate) [inline src/scripts/background/search-engines.ts:21]: I will update this later.

File: src/scripts/search-engines/qwant.ts [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+      '.ub-button': {
+        color: 'var(--ub-link-color, rgb(93, 96, 94))',
+      },
+      '.theme-dark .ub-button, [data-theme="dark"] .ub-button': {
+        color: 'var(--ub-link-color, rgb(122, 140, 109))',
+      },
+      '.ub-button:hover': {
+        textDecoration: 'underline',
+      },
+      'div[class*=WebResult-module__permalink]': {
```
Comment (iorate) [inline src/scripts/search-engines/qwant.ts:22]: `^=` may be better than `*=`. (Just a matter of taste?)

File: src/scripts/search-engines/qwant.ts [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+        actionTarget: '[class*=Text-module__permaLink]',
+        actionStyle: {
+          fontSize: '13px',
+          paddingLeft: '5px',
+        },
+      },
+      {
+        // Images
+        target: '[data-testid=imageResult]',
+        url: root => `https://${root.querySelector('cite')?.textContent || ''}`,
```
Comment (iorate) [inline src/scripts/search-engines/qwant.ts:83]: This is not the actual URL of the page, but we have to compromise.

File: src/scripts/search-engines/qwant.ts [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+        actionTarget: '.url',
+        actionStyle: {
+          fontSize: '13px',
+          paddingLeft: '5px',
+        },
+      },
+    ],
+    pagerHandlers: [
+      {
+        target: '[class*=Web-module], [class*=Stack-module__VerticalStack], div:empty',
```
Comment (iorate) [inline src/scripts/search-engines/qwant.ts:116]: Is `div:empty` necessary? I don't think such an element should have inner targets.
Please let me know if I miss anything.

**Review by iorate — CHANGES_REQUESTED** [review]
Would you check my comments?

File: src/scripts/search-engines/qwant.ts (hunk 7 above)
Comment (imagoiq) [reply src/scripts/search-engines/qwant.ts:22]: It will actually not work (that was my first attempt) as there are many classes and this one for example is the latest.

Comment (iorate) [reply src/scripts/search-engines/qwant.ts:22]: I see.

**Review by iorate — APPROVED** [review]
Thank you for your great work!

**Comment by imagoiq** [comment]
Almost… 🤔 you may have to revert, I've just found and issue, sorry. When navigating between web, news, images and videos, the control handlers are keeping elements from the current page and adding up to the amount of blocked sites.

**Comment by iorate** [comment]
I don't think it is a big problem. It is the expected behavior of "Solution 1" (although I may not have explained it enough).

Of course, if you are unsatisfied, you can modify it (take "Solution 2"). If so, would you please open a new pull request based on the current master branch?

**Comment by imagoiq** [comment]
You're right it's not a big problem, but soon or later, someone will complain. Solution 2 doesn't seem to fix everything but let's speak about it in the next PR.
